### PR TITLE
InstanceId TimeSeries

### PR DESCRIPTION
**Pull request by doctrino**
## Description
Please describe the change you have made.

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: .github/actions/setup/action.yml [hunk 1]
```diff
@@ -31,6 +31,6 @@ runs:
     - name: Install dependencies
       shell: bash
       run: |
-        pip install --upgrade pip poetry
+        python -m pip install --upgrade pip poetry
```
Comment (doctrino) [inline .github/actions/setup/action.yml:34]: The Windows failed without this one

File: cognite/client/data_classes/data_modeling/ids.py [hunk 2]
```diff
@@ -87,43 +90,26 @@ def load(
 
 
 @dataclass(frozen=True)
-class InstanceId:
```
Comment (doctrino) [inline cognite/client/data_classes/data_modeling/ids.py:90]: Moved it to avoid circular import.

File: cognite/client/_api/datapoints.py [hunk 3]
```diff
@@ -531,6 +541,7 @@ def retrieve(
         | DatapointsQuery
         | dict[str, Any]
         | SequenceNotStr[str | DatapointsQuery | dict[str, Any]] = None,
+        instance_id: None | InstanceId | Sequence[InstanceId] = None,
```
Comment (erlendvollset) [inline cognite/client/_api/datapoints.py:544]: These are only ever node ids.
```suggestion
        instance_id: None | NodeId | Sequence[NodeId] = None,
```

File: cognite/client/_api/datapoints.py [hunk 4]
```diff
@@ -1311,6 +1329,7 @@ def insert(
         | Sequence[tuple[int | float | datetime.datetime, int | float | str]],
         id: int | None = None,
         external_id: str | None = None,
+        instance_id: InstanceId | None = None,
```
Comment (erlendvollset) [inline cognite/client/_api/datapoints.py:1332]:
```suggestion
        instance_id: NodeId | None = None,
```

File: cognite/client/_api/datapoints.py [hunk 5]
```diff
@@ -1451,14 +1481,27 @@ def insert_multiple(self, datapoints: list[dict[str, str | int | list | Datapoin
...
+
+        DatapointsPoster(self, cdf_version).insert(datapoints)
 
     def delete_range(
         self,
         start: int | str | datetime.datetime,
         end: int | str | datetime.datetime,
         id: int | None = None,
         external_id: str | None = None,
+        instance_id: InstanceId | None = None,
```
Comment (erlendvollset) [inline cognite/client/_api/datapoints.py:1504]:
```suggestion
        instance_id: NodeId | None = None,
```

File: cognite/client/data_classes/datapoints.py [hunk 6]
```diff
@@ -165,6 +165,7 @@ class DatapointsQuery:
     )
     id: InitVar[int | None] = None
     external_id: InitVar[str | None] = None
+    instance_id: InitVar[InstanceId | None] = None
```
Comment (erlendvollset) [inline cognite/client/data_classes/datapoints.py:168]:
```suggestion
    instance_id: InitVar[NodeId | None] = None
```

File: cognite/client/data_classes/datapoints.py [hunk 7]
```diff
@@ -179,9 +180,9 @@ class DatapointsQuery:
     ignore_bad_datapoints: bool = _NOT_SET  # type: ignore [assignment]
     treat_uncertain_as_bad: bool = _NOT_SET  # type: ignore [assignment]
 
-    def __post_init__(self, id: int | None, external_id: str | None) -> None:
+    def __post_init__(self, id: int | None, external_id: str | None, instance_id: InstanceId | None) -> None:
```
Comment (erlendvollset) [inline cognite/client/data_classes/datapoints.py:183]:
```suggestion
    def __post_init__(self, id: int | None, external_id: str | None, instance_id: NodeId | None) -> None:
```

File: cognite/client/data_classes/datapoints.py [hunk 8]
```diff
@@ -488,6 +489,7 @@ def __init__(
         self,
         id: int | None = None,
         external_id: str | None = None,
+        instance_id: InstanceId | None = None,
```
Comment (erlendvollset) [inline cognite/client/data_classes/datapoints.py:492]:
```suggestion
        instance_id: NodeId | None = None,
```

File: cognite/client/data_classes/datapoints.py [hunk 9]
```diff
@@ -897,6 +905,7 @@ def __init__(
...
```
Comment (erlendvollset) [inline cognite/client/data_classes/datapoints.py:908]:
```suggestion
        instance_id: NodeId | None = None,
```

File: cognite/client/_api/datapoints.py [hunk 10]
```diff
@@ -1451,14 +1481,27 @@ def insert_multiple(self, datapoints: list[dict[str, str | int | list | Datapoin
...
+            raise ValueError("Input must be a list of dictionaries")
+        cdf_version: str | None = None
+        if any("instance_id" in d or "instanceId" in d for d in datapoints):
+            self._use_instance_api()
+            cdf_version = "alpha"
+            for d in datapoints:
+                if "instance_id" in d and isinstance(d["instance_id"], InstanceId):
+                    d["instance_id"] = d["instance_id"].dump(include_instance_type=False)  # type: ignore[assignment]
+                elif "instanceId" in d and isinstance(d["instanceId"], InstanceId):
+                    d["instanceId"] = d["instanceId"].dump(include_instance_type=False)  # type: ignore[assignment]
```
[the remark this answers is not in the document]
Comment (doctrino) [reply cognite/client/_api/datapoints.py:1495]: Actually, it is not necessary. It is already handled in `_verify_and_prepare_dps_objects` 🤦

File: cognite/client/_api/datapoints.py [hunk 11]
```diff
@@ -1615,18 +1667,23 @@ def insert(self, dps_object_lst: list[dict[str, Any]]) -> None:
...
                 dps_to_insert["externalId", xid].extend(validated_dps)
+            elif (instance_id := validated.get("instanceId")) is not None:
+                dps_to_insert["instanceId", (instance_id["space"], instance_id["externalId"])].extend(validated_dps)
             else:
                 dps_to_insert["id", validated["id"]].extend(validated_dps)
-        return list(dps_to_insert.items())
+        return [
+            ((id_name, {"space": id_[0], "externalId": id_[1]} if id_name == "instanceId" else id_), data)  # type: ignore[index, misc]
+            for (id_name, id_), data in dps_to_insert.items()
+        ]
```
[the remark this answers is not in the document]
Comment (doctrino) [reply cognite/client/_api/datapoints.py:1687]: Yes, you are right, simplified it. Note sure what made me go for this complex solution.

**Review by erlendvollset — APPROVED** [review]

**Pull request merged**